### ISSUE TITLE
Huge `import/no-cycle` performance downgrade in 2.25.3

### ISSUE THREAD

**Issue by zaicevas** [import-js/eslint-plugin-import#2348]
I upgraded `eslint-plugin-import` 2.22.1 -> 2.25.3 and the performance downgrade is huge. `eslint` is `7.24.0`. 

Roughly a hundred files:

```
Rule                                       |  Time (ms) | Relative
:------------------------------------------|-----------:|--------:
import/no-cycle                            | 304818.252 |    97.0%
prettier/prettier                          |   2347.669 |     0.7%
import/named                               |    754.553 |     0.2%
```

`2.24.2` is fine:

```
Rule                                       | Time (ms) | Relative
:------------------------------------------|----------:|--------:
import/no-cycle                            |  4206.707 |    23.0%
prettier/prettier                          |  1914.555 |    10.5%
react/prefer-stateless-function            |   849.317 |     4.6%
```

**Comment by ljharb** [comment]
Thanks, that's quite large.

The only changes that touched ExportMap in that range are 9a744f73 and #2212.

Do you have a TypeScript codebase? Also, can you confirm that v2.25.4 (which contains #2341) has the same speed drop?

**Comment by zaicevas** [comment]
> Do you have a TypeScript codebase? 

Yes

> can you confirm that v2.25.4 (which contains #2341) has the same speed drop?

Yes:

```
Rule                                          |  Time (ms) | Relative
:---------------------------------------------|-----------:|--------:
import/no-cycle                               | 300578.485 |    95.2%
prettier/prettier                             |   2384.299 |     0.8%
react/prefer-stateless-function               |    866.305 |     0.3%
```

**Comment by ljharb** [comment]
That strongly suggests it’s #2240. cc @mrmckeb

**Comment by mrmckeb** [comment]
Are you able to share a small reproduction of the issue @zaicevas? Or if not, can you share your `tsconfig.json`?

**Comment by zaicevas** [comment]
> Are you able to share a small reproduction of the issue @zaicevas?

Sorry, won't be able to do that due to lack of time.

`tsconfig.json`:
```
{
  "compilerOptions": {
    "target": "ES2016",
    "module": "commonjs",
    "jsx": "react-jsx",
    "noEmit": true,
    "strict": true,
    "noImplicitAny": false,
    "noImplicitThis": false,
    "baseUrl": "./app/",
    "esModuleInterop": true,
    "allowUnreachableCode": false,
    "isolatedModules": true
  },
  "include": [
    "./app/**/*.ts",
    "./app/**/*.tsx"
  ],
  "exclude": [
    "./node_modules/**/*"
  ]
}
```

**Comment by sadovnychyi** [comment]
Got 30s speedup (1:59s -> 1:29s for total linting time) by setting ignoreExternal to **false**.
```diff
-'import/no-cycle': [2, {ignoreExternal: true, maxDepth: 3}],
+'import/no-cycle': [2, {ignoreExternal: false, maxDepth: 3}],
```

**Comment by mrmckeb** [comment]
Thanks all, I did some quick testing around this today and can confirm that there is a massive performance downgrade between these releases.

### Findings

| Version  | New tsconfig resolution | Old tsconfig resolution |
|---|-------|----|
| 2.22.1 | ~80ms | ~80ms |
| 2.25.4 | ~230ms | ~230ms |

It seems that the issue is related to another change, @ljharb.

### Methodology

1. Install the latest `2.25.4` alongside a fresh Next.js app.
1. Copy this file in place of the version that shipped with `2.25.4`.
https://unpkg.com/eslint-plugin-import@2.22.1/lib/ExportMap.js
1. Modify the code (around line 705) to reimplement the new behaviour.
    ```js

          // const jsonText = _fs2.default.readFileSync(tsConfigInfo.tsConfigPath).toString();
          // if (!parseConfigFileTextToJson) {
          //   var _require = require('typescript');
          //   parseConfigFileTextToJson = _require.parseConfigFileTextToJson;
          // }
          // const tsConfig = parseConfigFileTextToJson(tsConfigInfo.tsConfigPath, jsonText).config;
          // return tsConfig.compilerOptions.esModuleInterop;

          const ts = require('typescript');
          const path = require('path');
          const configFile = ts.readConfigFile(tsConfigInfo.tsConfigPath, ts.sys.readFile);
          const config = ts.parseJsonConfigFileContent(
            configFile.config,
            ts.sys,
            path.dirname(tsConfigInfo.tsConfigPath),
          );
          // So we don't have to make other changes in this file.
          config.compilerOptions = config.options;
          return config;
    ```
1. Do the same in reverse (reimplement old behaviour in new version).

**Comment by zaicevas** [comment]
2.22.1 -> 2.24.2 is also a significant performance downgrade:

2.22.1:
```
import/no-cycle                            | 15407.126 |     6.1%
```

2.24.2:
```
import/no-cycle                   | 70428.732 |    36.3%
```

**Comment by zaicevas** [comment]
It turns out that `2.24.2` does not report `import/no-cycle` violations at all 😕 

```
// a.ts
import './b'

export function ab() {
  //
}

// b.ts
import { ab } from './a'
```

**Comment by DarkPurple141** [comment]
Keen to watch this issue, have noted this rule is eating into our CI, accounting for ~80% of our overall linting pipeline in a large-ish monorepo. I'm interested to know whether there is any thought on caching cycles previously searched across files to decrease the runtime performance hit of the rule.

**Comment by SDAdham** [comment]
same here, no-cycle seems to be a performance destroyer

**Comment by JounQin** [comment]
Can you try https://github.com/un-es/eslint-plugin-i to see whether it's much faster?

**Comment by SDAdham** [comment]
It didn't work:
```
eslint-plugin-import@npm:eslint-plugin-i: Registry failed to return tag "eslint-plugin-i"
```

**Comment by JounQin** [comment]
@SDAdham Sorry I don't use `yarn@v2+`, maybe you can find something similar to package aliasing.

**Comment by burtek** [comment]
> https://yarnpkg.com/features/protocols/#table
> 
> So it seems `eslint-plugin-import@npm:eslint-plugin-i@latest` should do the trick.

`eslint-plugin-import` is a nested dependency in my case, but it worked for me (using `yarn`) to use `eslint-plugin-i` via (in `package.json`):

```jsonc
{
  // rest of package.json
  "resolutions": {
    "eslint-plugin-import": "npm:eslint-plugin-i@latest"
  }
}
```

**Comment by dartess** [comment]
> Can you try https://github.com/un-es/eslint-plugin-i to see whether it's much faster?

Maybe I'm doing something wrong, but I got exactly the same result.

**Comment by dartess** [comment]
> Can you try https://github.com/un-es/eslint-plugin-i to see whether it's much faster?

Maybe I'm doing something wrong, but I got exactly the same result.

**Comment by lrdxg1** [comment]
Same here unfortunately. I was quite excited for this fix and was able to use the yarn resolution field to force `eslint-plugin-i` to be used but no difference.

Before:

```
Rule                                | Time (ms) | Relative
:-----------------------------------|----------:|--------:
import/no-cycle                     | 70282.194 |    38.1%
import/no-relative-packages         | 10151.240 |     5.5%
import/no-extraneous-dependencies   |  8517.592 |     4.6%
react/destructuring-assignment      |  7432.624 |     4.0%
import/order                        |  4776.704 |     2.6%
react/no-did-update-set-state       |  3923.661 |     2.1%
@typescript-eslint/no-unused-vars   |  3896.220 |     2.1%
react/no-unstable-nested-components |  2844.975 |     1.5%
react/no-deprecated                 |  2837.452 |     1.5%
import/extensions                   |  2737.678 |     1.5%
```

After:
```
Rule                                | Time (ms) | Relative
:-----------------------------------|----------:|--------:
import/no-cycle                     | 73077.763 |    37.5%
import/no-relative-packages         | 10611.396 |     5.4%
import/no-extraneous-dependencies   |  8764.613 |     4.5%
react/destructuring-assignment      |  7719.522 |     4.0%
import/order                        |  5061.809 |     2.6%
react/no-did-update-set-state       |  4245.396 |     2.2%
@typescript-eslint/no-unused-vars   |  4151.959 |     2.1%
react/no-deprecated                 |  3027.035 |     1.6%
react/no-unstable-nested-components |  2995.082 |     1.5%
react/function-component-definition |  2980.219 |     1.5%
```

**Comment by JounQin** [comment]
Then it's not a resolving issue but performance effects in `ExportMap`.

**Comment by evelant** [comment]
Anybody managed to solve this one? I just tried to add eslint to my repo for the first time ever and ran into this. Linting a single file takes 3+ minutes which is obviously not usable.

**Comment by MayGo** [comment]
Is this going to be fixed, or is there a workaround?

**Comment by ljharb** [comment]
@MayGo it will be fixed once there's a PR to fix it, and there's no workaround or else that would be the fix.

**Comment by tnguyen42** [comment]
Do we know if this issue exists on both npm and yarn, and if it exists on all versions of yarn? Or is there some other cases where it doesn't exist?
I'm surprised there is so little attention to it since it's been open for 15 months now

**Comment by ljharb** [comment]
@tnguyen42 npm or yarn should have no impact on it at all, since it's about runtime performance.

It's not that it's been paid "so little attention", it's that performance issues are very hard to reproduce reliably, and without that, they're very hard to fix.

**Comment by lukaselmer** [comment]
I found out a couple of things:

* Most time is spent within `ExportMap.parse`
* Reducing the depth improves the situation:
  * Depth 1: 3.92s
  * Depth 2: 10.40s
  * Depth 3: 14.46s
  * Depth 4: 15.56s
  * Depth 1000: 16.75s
* There is an option `ignoreExternal`
  * With this option on, the time goes down drastically to around 3s
  * However, when using absolute [imports / path mapping / aliases](https://www.typescriptlang.org/tsconfig#paths), it produces false positives
    * for this reason, `import/internal-regex` needs to be configured to include these files, for example:

```js
//  .eslintrc.js
module.exports = {
  settings: {
    'import/internal-regex': '^(@myNamespace/|client/|server/|shared/|__test__/|scripts/|types/)',
  },
  rules: {
    'import/no-cycle': ['error', { ignoreExternal: true }]
  },
  // [other config...]
}
```

* Most of the time spent in `ExportMap.parse` was used to parse files from libraries. The most impactful two libraries for me are
  * `@mui/icons-material`
  * `@mui/material`
* The main issue is IMO that the library reads and parses all the files
  * I first thought that `fs.readFileSync` may be the main issue (instead of  this,`import {readFile} from 'fs/promises'` and then `await fs.readFile(...)` could be used?), but according to the flame graph [1] and [2], parsing the file takes longer
  * Does it really have to parse all these files? From top to bottom? 🤔 
  * Here's a list of files which are read and parsed by my project [3]
    * When I ignore the `@mui` files, the lint time goes down drastically (from ~16s to ~4s), see [4]
    * I also checked if actually reading the file makes a difference, and it still takes around 4s when returning after reading the file [5]
* The sorted flamegraph looks like this: [6]
  * On the right side [7], we can see the garbage collection, and the `readFileSync` function, and the `visit` function
  * Then we see that a bit more to the left, `captureDependencyWithSpecifiers` spends most of the time reading files [8]
  * On the left side [9] is where the parsing happens
  * Maybe I don't see something obvious that someone else notices?
* Conclusion
  * Reproducing the performance issues should be pretty easy by installing `@mui-ions-material`, `@mui/material`, and creating a file which imports one or two things from these packages. Since I used my own project to do the analysis, I cannot guarantee it, but I'm pretty confident that it can be reproduced like this
  *  I don't know if it would help, but instead of doing everything in serial, we could use the `async` function to process / parse the files concurrently? There may be some IO that could be parallelized this way?
  * I don't understand why importing / analyzing external files makes sense - you should probably not have cycles between your code and external packages anyways, and if you do, you would notice it in `package.json` - right?
  * The analysis doesn't explain why updating eslint changes the performance of `import/no-cycle` - I'm just trying to understand why it's slow
  * For me, changing the configuration to ignore the external packages, and adding my own absolute prefixes, is a workable solution. Do you think it should be (better) documented that this is a potential solution if you're suffering from performance issues?

[1] Flamegraph overview:

<img width="1918" alt="image" src="https://user-images.githubusercontent.com/326935/232606214-7076bdb1-9157-457b-abd1-81c8e49acbd5.png">

[2]

<img width="606" alt="image" src="https://user-images.githubusercontent.com/326935/232606502-2b7eb01c-362d-472d-9614-e402434d3c3a.png">

<img width="618" alt="image" src="https://user-images.githubusercontent.com/326935/232606576-b21d24e3-54b2-4293-8b60-d0fc71dff15f.png">

<img width="2104" alt="image" src="https://user-images.githubusercontent.com/326935/232607286-bda7a962-3677-4835-adcb-50ddd715e1bc.png">

[3]

[parsedFiles.txt](https://github.com/import-js/eslint-plugin-import/files/11255723/parsedFiles.txt)

[4]

`ExportMap.js`

<img width="1494" alt="image" src="https://user-images.githubusercontent.com/326935/232611130-f364de0c-4df0-4d30-aa4b-0c7066bbe5b5.png">


<img width="946" alt="image" src="https://user-images.githubusercontent.com/326935/232611013-fc387da9-f1ae-489a-8300-ace08bfe3595.png">

[5]

<img width="1441" alt="image" src="https://user-images.githubusercontent.com/326935/232611714-a11171a6-062d-4132-b7ca-089d2f35aabc.png">


<img width="955" alt="image" src="https://user-images.githubusercontent.com/326935/232611675-c7aa0289-866f-4874-bc40-16d13d8007e5.png">

[6]

<img width="2467" alt="image" src="https://user-images.githubusercontent.com/326935/232613883-abf18022-1981-41e6-9f72-412eba5f3f04.png">

[7]

<img width="750" alt="image" src="https://user-images.githubusercontent.com/326935/232614562-a84461c1-520e-4de2-b9e5-c15c768b7572.png">

[8]

<img width="674" alt="image" src="https://user-images.githubusercontent.com/326935/232614716-bb84786c-3d33-4dce-9cac-22d45f2d4fe2.png">

[9]

<img width="2376" alt="image" src="https://user-images.githubusercontent.com/326935/232615717-efed2b54-c0b2-4528-9999-3aba5f451205.png">

**Comment by ljharb** [comment]
It has to parse every file fully in order to know what it imports/exports (or requires/module.exports, for CJS).

A half dozen rules depend on the export map - no-cycle is just one of them - but the map is cached across them, thankfully. This does mean that improvements in the export map benefit many rules at once.

I think it would be perfectly reasonable to detect the presence of `worker_threads` and do work in parallel when possible - the challenge would be to see if the communication overhead outweighed the CPU savings.

**Comment by lukaselmer** [comment]
Wow thanks for the quick reaction!

> It has to parse every file fully in order to know what it imports/exports (or requires/module.exports, for CJS).

Ok, I understand that. I thought that a "lightweight" import (which only scans for imports/exports/requires/module.exports) could be faster.

What about the external files? E.g. there's no chance that `@mui/material` imports something from my app.

> A half dozen rules depend on the export map - no-cycle is just one of them - but the map is cached across them, thankfully. This does mean that improvements in the export map benefit many rules at once.

Yes, I noticed, and I checked for that immediately, because it would have been a low-hanging fruit.

> I think it would be perfectly reasonable to detect the presence of worker_threads and do work in parallel when possible - the challenge would be to see if the communication overhead outweighed the CPU savings.

Yep. I'm not convinced it would bring any benefits, but I'll try implementing it in the next couple of days. Do you have a good tip on how to do performance tests? I'm currently modifying the compiled version directly in my sample app directory, and I'm thinking there must be a better way... :)

**Comment by ljharb** [comment]
Unfortunately I don't :-/ usually i `npm link` the plugin, and `npm link eslint-plugin-import` into the project i want to test it with, and then run `npm run build` in the plugin after I've made changes, and then rerun the linting task in the project.

**Comment by TkDodo** [comment]
> Got 30s speedup (1:59s -> 1:29s for total linting time) by setting ignoreExternal to **false**.
> 
> ```diff
> -'import/no-cycle': [2, {ignoreExternal: true, maxDepth: 3}],
> +'import/no-cycle': [2, {ignoreExternal: false, maxDepth: 3}],
> ```

I can confirm this. We had `ignoreExternal` on, so linting took around ~3 minutes. After I turned it off, it went down to ~1 minute

**Comment by ljharb** [comment]
I'd expect an "ignore" option to make things *faster*, not slower, so there must be some low-hanging fruit here - a PR would be welcome.

**Comment by XDean** [comment]
Is there any solution or plan? The newest 2.27.5 still very slow.

**Comment by sztadii** [comment]
> Got 30s speedup (1:59s -> 1:29s for total linting time) by setting ignoreExternal to **false**.
> 
> ```diff
> -'import/no-cycle': [2, {ignoreExternal: true, maxDepth: 3}],
> +'import/no-cycle': [2, {ignoreExternal: false, maxDepth: 3}],
> ```

Also I can confirm. From 13 minutes to 2 min.

**Comment by soryy708** [comment]
Here's an idea for how to make `import/no-cycle` faster:
https://github.com/import-js/eslint-plugin-import/issues/2937

**Comment by martin-mindflow** [comment]
Bumping this, it's a also a big problem in our (semi-large) monorepo, with `no-cycle` accounting for ~80% of the relative linting time. None of the workarounds of this thread worked for us :( so far we've simply resorted to outsourcing cycle detection to another tool (madge), but the DX is clearly worse than having it build-in to eslint

**Comment by diegodorado** [comment]
Same issue here: with `ignoreExternal` it is actually worst... I was expecting that flag to improve linting time

**Comment by ljharb** [comment]
When #2998 is merged and released, this should improve drastically.

**Comment by jzaefferer** [comment]
@ljharb thanks for getting that merged! Any estimate when it could get released? I'm mostly asking since the last regular release was back in December, so it's hard to guess when the next one might come around.

**Comment by soryy708** [comment]
> @ljharb thanks for getting that merged! Any estimate when it could get released? I'm mostly asking since the last regular release was back in December, so it's hard to guess when the next one might come around.

The maintainer is working on flag config support, so after it will be done, expect a minor version I think?

**Comment by soryy708** [comment]
@mrmckeb regarding previous conversation about your contribution: https://github.com/import-js/eslint-plugin-import/pull/2240

Resolving TypeScript config more correctly is a good change, but I'm curious if it could be a big contributor to filesystem accesses reported in profiling?

- I don't know how many FS reads TS needs to compile an extended TS config. Can this be profiled?
- Is it reasonable to put extended ts config behind a feature flag (lint rule option)? so only those who have multiple tsconfigs in hierarchy will be impacted by the performance hit?
- Is it possible to cache results instead of resolving it from scratch each time?

**Comment by mrmckeb** [comment]
@soryy708, as noted here, I did test this and found that it wasn't the cause of the significant performance impact noted in this thread.
https://github.com/import-js/eslint-plugin-import/issues/2348#issuecomment-1006299367

I didn't go into profiling, so I'm not sure of the exact performance impact.

I think in-memory caching of the fully resolved/parsed config is a good idea.

Putting extended configs behind a flag is probably the wrong solution here, especially as we don't know if it has caused any performance issues:
- If the `extends` field isn't there, there shouldn't be a lookup (this is using native TypeScript functionality).
- If the `extends` field is there and not being resolved, `eslint-plugin-import` won't work as expected.

**Comment by soryy708** [comment]
The SCC optimization has been released, try v `>=2.30.0`

**Comment by ljharb** [comment]
I'm going to close this; if there's still a perf issue in v2.30+, please file a new issue.
